### ISSUE TITLE
Broken test

### ISSUE THREAD

**Issue by albop** [levinboimtomer/SPG#1]
On a fresh 64 bit anaconda (Python 2.7, both Linux and Windows) I get the following output:

```
 Iteration   FunEvals Projections     Step Length    Function Val        Opt Cond
        1         82          3     8.14623e-04     1.87314e+02     7.72915e+01
        2        123          5     1.00000e+00     6.10225e+01     2.34484e+01
        3        164          7     1.00000e+00     3.27345e+01     1.96349e+01
        4        205          9     1.00000e+00     2.09821e+01     1.14803e+01
        5        246         11     1.00000e+00     1.45962e+01     1.13655e+01
        6        287         13     1.00000e+00     1.19849e+01     1.18290e+01
        7        328         15     1.00000e+00     8.34070e+00     6.63697e+00
        8        369         17     1.00000e+00     6.79721e+00     4.82917e+00
        9        410         19     1.00000e+00     5.60566e+00     4.05898e+00
       10        451         21     1.00000e+00     2.95021e+00     3.71048e+00
       11        492         23     1.00000e+00     1.98367e+01     3.40016e+01
       12        533         25     1.00000e+00     7.11594e+00     1.22811e+01
```

Since the step length is always 1, there may be something wrong with the line search. @levinboimtomer , have you tested it recently ?

**Comment by levinboimtomer** [comment]
This is for the provided test case?
Otherwise, what is your objective function and gradient?

**Comment by albop** [comment]
Yes, I just ran `python SPG_test.py` . Then I noticed, that the `my_test` has a quasi-option `test_mode`  so I changed it to `1` and ran a similar result instead of the one in comments.

**Comment by levinboimtomer** [comment]
I just ran it with test_mode == 1 and got similar results to the ones in the comments (see trace below).
What is the optimality condition ("Opt Cond") you get when test_mode is set to 1? 
is it around default_options.optTol (default is 1e-5).

When I set test_mode to 0, I don't get convergence within 500 iterations,
in this case, increasing the number of iterations to 50,000, I did get convergence.

There are two things to note
1. default_options.maxIter (set to 500 by default) actually counts the number of function evaluations and so, does not correspond directly to the number of gradient steps taken.
2. Spectral GD may increase the function value in some iterations.
You can shut it off with 
default_options.useSpectral = False

==== test mode = 1 results
         1          8          3     2.83106e-02     9.82156e+00     1.28787e+01
         2         12          5     1.00000e+00     6.72665e+00     1.05669e+01
         3         16          7     1.00000e+00     4.96280e+00     6.26491e+00
         4         20          9     1.00000e+00     4.35856e+00     4.59714e+00
         5         24         11     1.00000e+00     3.46092e+00     2.84511e+00
         6         28         13     1.00000e+00     3.77708e+00     8.62512e+00
         7         32         15     1.00000e+00     4.38154e+00     3.07368e+00
         8         36         17     1.00000e+00     3.11077e+00     1.55083e-01
         9         40         19     1.00000e+00     3.11036e+00     1.38575e-01
        10         44         21     1.00000e+00     3.10915e+00     1.84073e-02
        11         48         23     1.00000e+00     3.10914e+00     1.72484e-02
        12         52         25     1.00000e+00     3.10913e+00     1.41742e-03
        13         56         27     1.00000e+00     3.10913e+00     7.24786e-04
        14         60         29     1.00000e+00     3.10913e+00     1.80905e-03
        15         64         31     1.00000e+00     3.10913e+00     2.90637e-05

**Comment by albop** [comment]
Yes, `test_mode=1`  does converge. I misread the figures. They are indeed the expected ones.
As for the other example, with maxIter=60000, optimization ends with: 

```
      1044      54243       2367     1.00000e+00     2.21996e-05     5.89424e-04
      1045      54284       2369     1.00000e+00     2.20747e-05     8.10256e-04
      1046      54407       2373     3.64889e-02     2.20611e-05     1.66137e-03
      1047      54448       2375     1.00000e+00     2.20602e-05     2.00138e-03
Function value changing by less than progTol
```

Perdon my ignorance, but what does the optimality criterion mean in this case ? Is there a guarantee that we have a local minimum ?
Also I'm confused by the step_length output. Apparently it is not the distance between two successive guesses. Then what is it ?

I tried your suggestion of turning off spectral GD. It now takes ~300000 evaluations instead of ~50000 and the solutions are relatively close. But it seems the function values are still non monotonous:

```
      1476     198276       6312     2.16773e-01     1.26437e-03     5.47610e-02
      1477     198399       6316     7.34386e-03     1.22161e-03     1.66561e-02
      1478     198522       6320     3.31947e-02     1.27482e-03     6.05393e-02
      1479     198645       6324     6.24604e-03     1.21480e-03     6.58464e-03
```

**Comment by levinboimtomer** [comment]
optCond captures how much the current iterate x_i changed from the previous one x_{i-1}
Optimization stops when it is below a threshold (options.optTol), 
There's another threshold, optProg, which stops optimization if the gradient*stepsize is too small, or if the function value stopped changing.
The code does not provide mathematical guarantees of convergence, such as duality gap, but I have used it successfully in several projects.

SPG to regular Projected Gradient Descent (PGD):
Try setting options.memory = 1 (keeping useSpectral = False),
or, change line 108 from,
  if options.memory == 1: 
to 
  if options.memory == 1 or not useSpectral: 

While developing, its good to start with PGD and check the gradients are correct (that is, you see convergence).
Then, if optimization takes too long, try turning SPG back on, and play with the size of memory (usually I set it to 5).
In practice, I don't always see faster convergence with SPG.
